### DISCUSSION TITLE
Using the Blockified Single Product Template

### DISCUSSION THREAD

**Discussion by sunyatasattva** [woocommerce/woocommerce#44953] [WooCommerce, Blocks, and Site Editing]
Recently, I have tried implementing a real-world design for a Single Product using our latest blocks. This discussion is about some of the road-blocks I have encountered and about potential ways to move forward.

The main issues I have encountered stemmed from a main limitation: **The Product Details block contained three panels: “Description”, “Additional information” and “Reviews”. For my design, I needed these three tabs to be separated.**

Since the “Description” is basically the “Post Content”, I thought I could use the core block there, and that's when I discovered some weird incompatibility. By placing the “Post Content” block in my template, the “Product Details” block would be gone. 😨 

The reason for this is that the ”Post Content” block [calls `the_post` function](https://github.com/WordPress/gutenberg/blob/f7052d3e1e3e69f3b09b412c071c6c1eac69836f/packages/block-library/src/post-content/index.php#L41), which is a function with [global side-effects](https://developer.wordpress.org/reference/functions/the_post/): it doesn't just set the context, but in fact moves the index of the loop forward when it is called.

Looking at how our “Product Details” block work, because the logic was copied from the WooCommerce classic template, [it uses the loop and also calls `the_post` function](https://vscode.dev/github/woocommerce/woocommerce-gutenberg-products-block/blob/363f904000438765b08e34b6947accc73232e5d7/src/BlockTypes/ProductDetails.php#L54-L57). This means that the two conflict as the order in which the blocks have been laid out within the page will matter and disrupt the rest of the blocks.

## Possible ways forward

In general, having blocks causing global side-effects is in my opinion ill-advised. However, it seems that the ”Post Content” block decision was made to maintain backwards compatibility. According to the comments above that aforementioned Gutenberg code, it seems that the only backwards compatibility which is needed is to set `in_the_loop` to `true`. So, I wonder if we could perhaps change that code upstream to figure out how to minimize any side effects. Paging @mamaduka and @aristath, as they were the one introducing that change, just for context.

In any case, changing Gutenberg's code is not a high priority as it wouldn't change the fact that we should avoid side effects in our own code if we want, e.g., the “Description” and the “Reviews” block to be separated.

We have a few ways of approaching this which I can think of, considering the postulate that We want to be able to reuse existing templates in order to maintain a layer of backwards compatibility (run all the hooks etc.) and also provide the fastest route to making the templates FSE-compatible (please correct me if this postulate is wrong).

### Override global variables that the templates need

The templates, such as the [single-product-reviews](https://github.dev/woocommerce/woocommerce/blob/591921cd30ebcb5a21ef4006f5db8da3b96ab123/plugins/woocommerce/templates/single-product-reviews.php), rely on global variables to be set, which is why the first approach was to copy the old logic and making sure that everything that the loop provides is set.

Our blocks could make sure these variables are provided by hard replacing them, like so:

```php
protected function render( $attributes, $content, $block ) {
  $post_id = $block->context['postId'];
  $old_global_product = $GLOBALS[ 'product' ];

  // Overwriting the global, because we need it in the template
  $GLOBALS[ 'product' ] = wc_get_product( $post_id );

  wc_get_template( 'single-product-reviews.php' );

  $reviews = ob_get_clean();

  // Resetting the global to what it was
  $GLOBALS[ 'product' ] = $old_global_product;
```

This is not the cleanest approach and, while we reset the global to the old one, there is still some risk of side effects (though relatively small, as we only render the template). But it's the fastest solution: we figure out all the globals that are required by a template, and we pass them on. I think it's just `$product` and the global `$query` object.

### Pass the variables to the template when calling it

The `wc_get_template` function allows you to pass an array of additional data or variables that can be used within the template file you’re loading as a second argument.

However, as the templates are explicitly declaring the use of the `global` variable instead of any locally scoped variable with the same name. So this will take precedence.

We could modify the WooCommerce template upstream to allow to accept those variables, like so:

```php
<?php
/**
 * Display single product reviews (comments)
 *
 * This template can be overridden by copying it to yourtheme/woocommerce/single-product-reviews.php.
 * @see     https://docs.woocommerce.com/document/template-structure/
 * @package WooCommerce\Templates
 * @version 4.3.0
 */

defined( 'ABSPATH' ) || exit;

// Use product from $args if available, otherwise fall back to global $product
$product = isset($product) ? $product : $GLOBALS['product'];

// ...
```

This would not change the behavior of the template and is a minimally invasive intervention which also has the nice side-effects of allowing templates to be basically reusable in any context, by just passing the correct context variables.

### Override the core templates with our own

As far as I know, `wc_get_template` **function** will take templates overrides from the current theme, but not from plugins. We can, however, use the `wc_get_template` **filter** to hook there and pass our own template overrides, in which we can decide how to deal with these issues (e.g. using the approach above, without changing the core template).

I'm actually not sure what advantage this approach would have, besides being faster than the second one (as we don't need to get anything merged in core) and giving us control on how we want to deal with FSE-related issues.

However, I personally don't consider these advantages interesting enough for the drawback of having a custom implementation. Especially now that we are part of the Core Product Team, I think the second solution seems more attractive than this one at least.
 
---

So, this is a rundown of the problem and the proposed solutions. Paging @gigitux as the DRI of the project, @nefeline and @thealexandrelara as the developers, and @nerrad to have a broader overview on the direction. Specifically, the action points we'd need to get out of this discussion are:

- [x] Propose any other solutions besides the ones I have considered.
- [x] Decide which is the way forward among the possible solutions.
- [x] Decide who will take ownership of the solution, and what would the time frame be.

Just FYI, I have already implemented the full “Reviews” block using the Solution 1, as I'm going to need something relatively quickly for the project I'm working on. If this is the way to go, I can open a PR. Otherwise, we can think of a longer term solution together.

**Comment by ntsekouras** [comment]
I'm not aware of the nuances here, but my first thoughts were more about whether we can pass arguments through context(solution 2). 

Do the new Woo blocks use the postIds from context or the rely on the `loop` and its globals? If they do rely on the globals, can they be updated?

If I remember correctly the issue was that some plugins(including Woo) had some checks for `$wp_query->in_the_loop` in some templates and that value was `false`, making the content not render. Is this still the case?

**Reply by sunyatasattva** [reply]
Hey thanks as usual for chiming in! I guess we need to distinguish here two things: in general, new Woo blocks do use `postId` from context and do not rely on the loop. However, in the context of blockifying the Single Product Template we made some decisions, in order to both be faster to produce something and also maintaining backwards compatibility, and certain parts basically all they do is call the relevant legacy template part.

In other words, our very first step was to create the “Classic templates”, which were monolithic blocks which encompassed whole templates, just so they would be minimally compatible with FSE, allowing merchants to customize everything that was _around_ the template.

The second step (which we are currently in), consisted of breaking down this monolithic template into smaller pieces and have each block be responsible for rendering a piece of the legacy template (e.g. the “Add to Cart form” or the “Product Details”).

Now, please @gigitux correct me if I'm wrong but I don't think we necessarily want to transition away from this approach anytime soon. We are still going to use the smaller templates because they are (a) quick wins and (b) still render all the filters and hooks that 3PDs would be expecting (and also benefit from theme override). In my case, I'm now needing to break down the template even further, as the “Product Details” block includes “Description”, “Additional Details“ and “Reviews” and I'll need to separate them, while still rendering the core templates for those blocks.

> I'm not aware of the nuances here, but my first thoughts were more about whether we can pass arguments through context(solution 2).

I'm also for solution 2, for the record. Those templates still use not only the `global $product` but also rely on some loop functions, and I'm not too familiar. I think we should still be able to pass everything through context (basically, we need to replace the `global $wp_query` if passed), but I'm not 100% sure and perhaps someone could provide confirmation, if they are aware.

> If I remember correctly the issue was that some plugins(including Woo) had some checks for $wp_query->in_the_loop in some templates and that value was false, making the content not render. Is this still the case?

This seemed to be the reason while reading the PRs that introduced those changes. I can't answer right now, I wonder if @gigitux knows? In any case, would it perhaps make more sense to change those blocks to set `$wp_query→in_the_loop` to `true` instead of calling `the_post` or other functions which mess with the current index of the loop?

It's not blocking for me now, but I still think that talking about the side-effects of those blocks is probably wise at this point.

**Comment by gigitux** [comment]
Thanks, @sunyatasattva! I agree with @ntsekouras about the solution that we should follow. @thealexandrelara and I noticed a similar issue while we worked on the Single Product Block. For this reason, we refactor the Product Add to Cart form block [in this way](https://github.com/woocommerce/woocommerce-blocks/pull/8610/files#diff-b6dbf51b98c926a1ad3723993ddcb154b608d8abdda4d6282e38d325aa4f5ea8R28-R48).

Also, for the variations (Product Title -> variation of Post Title and Product Summary -> variation of Post Excerpt), where we can't edit the render function, we replace the global variable, and after that, we restore the original one. ([source code](https://github.com/woocommerce/woocommerce-blocks/blob/b97bd02371f9ec91359b2b361888a9e4a2b35c27/src/BlockTypes/SingleProduct.php/#L76-L89)).

I didn't have time to open an issue, but this is necessary because in WP 6.2, both blocks use a function that uses the global variable:
- Post Title ([source code](https://github.com/WordPress/gutenberg/pull/48001/files#diff-7b855403e27d2c801f7c373431522a9ec3326befb3bc3929c8ed725e58b30e5fL22))
- Post Excerpt ([source code](https://github.com/WordPress/gutenberg/blob/b93fd88fbca8c701a4d017d71a3e07dfc5318ea9/packages/block-library/src/post-excerpt/index.php/#L31-L36))


For the Post Title, in the last month, [a PR fixed this behavior](https://github.com/WordPress/gutenberg/pull/48001).

**Reply by sunyatasattva** [reply]
Hey @gigitux!

Thanks for the perspective.

> Thanks, @sunyatasattva! I agree with @ntsekouras about the solution that we should follow. @thealexandrelara and I noticed a similar issue while we worked on the Single Product Block. For this reason, we refactor the Product Add to Cart form block [in this way](https://github.com/woocommerce/woocommerce-blocks/pull/8610/files#diff-b6dbf51b98c926a1ad3723993ddcb154b608d8abdda4d6282e38d325aa4f5ea8R28-R48).

I'm not sure we are talking about the same solution here, however. Because the solution you have implemented sounds more like solution 1, than solution 2. However, the second link seems to point still to the Product Add to Cart form and not the variations you mention, so maybe I misunderstood?

And the “Product Add to Cart” form, while I see you have removed reliance on the loop, it doesn't seem it suffers from the same problem as calling `do_action( 'woocommerce_' . $product->get_type() . '_add_to_cart' );` is all you need. I see that you do get the `$product` independently, which is part of what I'm talking about, but you are not passing a specific context to a PHP template file, as I would need to do with the `single-product-reviews.php` template file.

Again, I'm not too familiar with these things, so if I'm missing something please let me know. We can also pair when needed. To me, it seems they are two slightly different problems.

I still agree that the best solution should be the second one I proposed. I requires modifications in WooCore though.

**Reply by gigitux** [reply]
Sorry, I misunderstood 😓

> I'm not sure we are talking about the same solution here, however. Because the solution you have implemented sounds more like solution 1, than solution 2. However, the second link seems to point still to the Product Add to Cart form and not the variations you mention, so maybe I misunderstood?

You are right!

> And the “Product Add to Cart” form, while I see you have removed reliance on the loop, it doesn't seem it suffers from the same problem as calling do_action( 'woocommerce_' . $product->get_type() . '_add_to_cart' ); is all you need. I see that you do get the $product independently, which is part of what I'm talking about, but you are not passing a specific context to a PHP template file, as I would need to do with the single-product-reviews.php template file.

Good point! Yep, you are right. Not the best example :/ 
A better example is the Product Gallery Image (https://github.com/woocommerce/woocommerce-blocks/blob/b97bd02371f9ec91359b2b361888a9e4a2b35c27/src/BlockTypes/ProductImageGallery.php/#L41-L64) where we used the solution 1.

> it seems they are two slightly different problems.

Are you referring to the variations? Yep, thinking about it, you are right!

**Reply by sunyatasattva** [reply]
> A better example is the Product Gallery Image (https://github.com/woocommerce/woocommerce-blocks/blob/b97bd02371f9ec91359b2b361888a9e4a2b35c27/src/BlockTypes/ProductImageGallery.php/#L41-L64) where we used the solution 1.

Just a heads up, it seems that in your code sample you are not restoring the `global` to its original value. It's probably safer to do so, if we decide to go ahead with these kinds of solutions.

**Comment by nefeline** [comment]
Hi there, folks!

Thanks @sunyatasattva, for this great discussion and for sharing all the details about your investigation. I have a different approach to propose: rely on `rewind_posts();` instead.

Sharing a few more details:

The global `$wp_query` variable holds an array of posts. When we loop through this array using the while `have_posts()` loop, the internal pointer is advanced to the next position in the array with each iteration.

However, in our case here, we need to loop through the same set of posts again, and this is where `rewind_posts()` comes in handy. This function resets the internal pointer of the $posts array to the beginning, allowing us to loop through it again.

To exemplify:

- When adding the `$wp_query` global variable within the `render_tabs` function for the [Product Details block](https://github.com/woocommerce/woocommerce-blocks/blob/90957206ec7085443510da322e1d01faa34b8e4e/src/BlockTypes/ProductDetails.php#L54), it is possible to visualize the following:

<img width="1241" alt="Screenshot 2023-03-21 at 20 40 19" src="https://user-images.githubusercontent.com/15730971/226766188-06ac5817-59d6-4c17-b9a9-090d3bf3e544.png">

Notice how `current_post` is already zero: in other words, the loop kicked off elsewhere and was already finished by the time it reached the Product Details block.

To circumvent this, we can simply reset the internal pointer of `$wp_query->current_post` to the beginning of the array via `rewind_posts`:

```php
	protected function render_tabs() {
		ob_start();

		rewind_posts();
		while ( have_posts() ) {
			the_post();
			woocommerce_output_product_data_tabs();
		}

		return ob_get_clean();
	}
```
 Now, with the `$wp_query->current_post` value set to `-1`, the while loop with `have_posts()` can successfully kick-off:

<img width="1239" alt="Screenshot 2023-03-21 at 21 33 35" src="https://user-images.githubusercontent.com/15730971/226772208-17106bd5-fbdb-42df-8d6a-cc595ba210dd.png">

### Reasons to follow this approach:

In WordPress, the `while ( have_posts() ) { the_post(); }` loop is used to iterate through a list of posts in the current query and display their content.

Even though the single-product template is intended to display only one product, it still uses this loop to retrieve and display the content of the current post.

This is because the `have_posts()` function checks if any posts are available in the current query, and in the case of the single product template, the query is set up to retrieve only one - the current product being viewed.

Next, the `the_post()` function sets up the global $post variable with the current post being viewed, which allows us to access and display its content using WordPress template tags and functions.

So, even though there is only one post (aka product) being displayed in the single product template, the `while ( have_posts() ) { the_post(); }` loop is still used to set up the necessary post data and display the content in a consistent and standardized way. 

For this reason, I believe it is very important for us to keep relying on the [native loop mechanism](https://developer.wordpress.org/themes/basics/the-loop/#individual-post) that WordPress uses for outputting posts through a template and not diverge from that.

With this approach, no upstream changes are required to the core of Woo or Gutenberg: `rewind_posts();` already takes care of everything we need.

### Sidenote

Regarding:

> @thealexandrelara and I noticed a similar issue while we worked on the Single Product Block. For this reason, we refactor the Product Add to Cart form block [in this way](https://github.com/woocommerce/woocommerce-blocks/pull/8610/files#diff-b6dbf51b98c926a1ad3723993ddcb154b608d8abdda4d6282e38d325aa4f5ea8R28-R48).

@gigitux @thealexandrelara I'm adding a review regarding [this change](https://github.com/woocommerce/woocommerce-blocks/pull/8610/files#diff-b6dbf51b98c926a1ad3723993ddcb154b608d8abdda4d6282e38d325aa4f5ea8R28-R48): unsure about the reasoning here, but I wouldn't recommend making this change for the same reasons shared here earlier.

### Final notes

Regarding:

> Decide who will take ownership of the solution, and what would the time frame be.

I'm glad to take ownership and open a PR with [this fix](https://github.com/woocommerce/woocommerce-blocks/commit/26bb68a2ec32e8c511f57f3ab5790f51a8e63bf5) if the team agrees and no one has any concerns with the proposed approach. cc: @danielwrobert

**Reply by sunyatasattva** [reply]
Hey @nefeline! As usual, thank you for your extremely detailed run-down, also TIL about `rewind_posts`!

> Notice how current_post is already zero: in other words, the loop kicked off elsewhere and was already finished by the time it reached the Product Details block.

Yes, this is what I explain in the OP:

> The reason for this is that the ”Post Content” block [calls the_post function](https://github.com/WordPress/gutenberg/blob/f7052d3e1e3e69f3b09b412c071c6c1eac69836f/packages/block-library/src/post-content/index.php#L41), which is a function with [global side-effects](https://developer.wordpress.org/reference/functions/the_post/): it doesn't just set the context, but in fact moves the index of the loop forward when it is called.

I have to say that I disagree with your proposed approach: even though rewinding the post pointer would fix it, it is liable to cause unpredictable side-effects down the line.

> So, even though there is only one post (aka product) being displayed in the single product template, the while ( have_posts() ) { the_post(); } loop is still used to set up the necessary post data and display the content in a consistent and standardized way.
> 
> For this reason, I believe it is very important for us to keep relying on the [native loop mechanism](https://developer.wordpress.org/themes/basics/the-loop/#individual-post) that WordPress uses for outputting posts through a template and not diverge from that.

Blocks are, by their nature, modular and composable. That's entirely different from the classic WordPress approach. I don't think that, conceptually, blocks should cause global side-effects. If we rewind the posts, we might be breaking another block which was relying on the counter being incremented by the “Post Content” block. It's just an arms race, pretty much.

Also, besides the “Post Content” block which uses the native loop for the reasons that @ntsekouras described above, I am not aware of any other block which uses The Loop in this manner. All the blocks pretty much work by receiving either attributes by the users or retrieving it via `usesContext` in the `block.json`, so I don't think that it would be in line with the current design philosophy, and we would thus be diverging from that.

Furthermore, I believe it would be very unwieldy if every element block (title, author, comments, rating, etc.) would have to use the loop and rewind it.

As always, happy to change my mind if proven wrong, or just agree to disagree if the majority thinks your proposed solution is the way forward.

**Reply by nefeline** [reply]
Thanks for the input @sunyatasattva !

> I have to say that I disagree with your proposed approach: even though rewinding the post pointer would fix it, it is liable to cause unpredictable side-effects down the line.
> If we rewind the posts, we might be breaking another block which was relying on the counter being incremented by the “Post Content” block. It's just an arms race, pretty much.

I think this would be a concern if we were talking about the main blog loop, no?  For an individual post, the array of posts is always one, so I can't think of anything that could rely on incrementing the “Post Content” block or any other.

> Blocks are, by their nature, modular and composable. That's entirely different from the classic WordPress approach. I don't think that, conceptually, blocks should cause global side-effects. If we rewind the posts, we might be breaking another block which was relying on the counter being incremented by the “Post Content” block. It's just an arms race, pretty much.

Both the "Post Content" and the "Post Featured Image" blocks already do something along these lines on Gutenberg by invoking `the_post()`, so I'm not sure why it would be a problem for us to follow the same path: I'm struggling to think about a scenario where this could be a problem so an example here would be helpful. 

The reason why I'm more inclined to keep `( have_posts() ) { the_post(); }` is, in addition to the reasons mentioned earlier, also to ensure we are 100% aligned with what the core of Woo does now (no hooks skipped): furthermore, 

I suspect _not_ having it in place has the potential to cause inconsistencies down the line: maybe one alternative could also be to place it (or the hooks associated with it) on the compatibility layer instead so multiple blocks could share the same loop if needed?

If we bypass `have_posts()`, what will signal the loop started/ended for that particular product? What about plugin developers relying on `'loop_end'` or `'loop_no_results'`? Are we ok with not giving support to those?

I'm happy to change my mind if folks are more inclined to stick with option 2, although I'm a bit concerned with the possible side effects of this approach as well.

**Reply by thealexandrelara** [reply]
I'm also inclined to solution 2 since, as @sunyatasattva  mentioned is minimally invasive and would provide a lot of flexibility without us having to reimplement everything on our side (which also extends to other projects). 

We had to rely on something really similar to the 1st solution for the Single Product block implementation because since we are using variations, we cannot edit the `render` function, and they are using the global post object to display the data, but the way we did it was definitely not the cleanest since we had to loop over each block inside the Single Product Block and replace the global post

> @gigitux @thealexandrelara I'm adding a review regarding [this change](https://github.com/woocommerce/woocommerce-blocks/pull/8610/files#diff-b6dbf51b98c926a1ad3723993ddcb154b608d8abdda4d6282e38d325aa4f5ea8R28-R48): unsure about the reasoning here, but I wouldn't recommend making this change for the same reasons shared here earlier.

Thank you for such a complete answer to this discussion! 

The change we made was necessary because the initial approach for the Add to Cart Form block was considering that the merchant would include the block only inside the Single Product Template. The Single Product Template works well because when the user accesses the Product page, the loop is invoked and provides access to the post data for the current Product that the user is accessing.

When we started creating the Single Product block, we needed to reuse the completed blocks; one of those blocks is the Add to Cart Form. The merchant can add the Single Product block in basically anywhere (inside a Post, a Page, a template other than the Single Product Template, etc), and the merchant can select any product they want, so the `productId` of the selected product will be saved in the database.

If we rely on invoking `the_post`, we would display information for the current post that the user is in and not from the product selected on the Editor. So we had to make that change in order to retrieve the `productId` from the Context/attributes and fetch the correct product data. Considering that, do you know if we can still rely on the loop to fetch information for a specific product or it would be a limitation of this approach?

**Reply by sunyatasattva** [reply]
Thank you both for your inputs. @nefeline I hear your points: preserving the loop would allow to preserve all the legacy logic associated with it. Also, thank you for providing another example in the “Featured Image” block, because it made me realize that maybe I'm missing something:

If both blocks are present on the same page, since they do not use the `rewind_posts` mechanism that you proposed, wouldn't calling `the_post` increase the counter and make them out of sync? Then, maybe they have some other way to keep the loop from making side effects (which is my main worry)?

If I have misunderstood the loop and there is indeed a way to call it side-effects free, I would be much more inclined in agreeing with you fully. As it stands, I don't think that the composable nature of blocks is very much in line with those kind of global side effects.

Perhaps this would be one of those instances in which someone who is more aligned with the overall direction and philosophy of Gutenberg could chime in, so we don't take a decision which is not in line with that (which I think I understand it's also your concern).

From what I've gathered from @ntsekouras's comment, it seems that those instances are more the exception than the rule and should be avoided. But perhaps he could confirm, or we could get a take from @danielwrobert, @tjcafferkey or @nerrad.

In the end, like you say, I'm happy to do what is consistent with everything else.

**Reply by nerrad** [reply]
> If both blocks are present on the same page, since they do not use the rewind_posts mechanism that you proposed, wouldn't calling the_post increase the counter and make them out of sync? Then, maybe they have some other way to keep the loop from making side effects (which is my main worry)?

I think this should be checked. In the case of _single products (post)_ loops I think the concerns are less here than if you were looking at a loop returning multiple posts. 

I appreciate the concern around side effects, however, this is more of a case of _controlling_ the side-effects to minimize the impact vs trying to completely eliminate them. Especially as we optimize towards preserving compatibility at this point.

**Reply by sunyatasattva** [reply]
I've made [a PR](https://github.com/woocommerce/woocommerce-blocks/pull/8857) with my proposed solution (1.) from the OP. I did this because it was the lowest hanging fruit for now and we can always change the implementation details transparently when we find out a better and cleaner way (I still would vote for solution 2., personally).

I've tried to implement @nefeline's proposed solution and instead of replacing the global `$wp_query` I had changed my code like so:

```php
rewind_posts();
if ( have_posts() ) {
  the_post();
}
```

This still conflicted with the Post Content block, and I didn't dig any deeper to be honest. Again, perhaps we can change these details later to unblock this issue? In any case, this solution I have implemented still uses the native WordPress loop mechanism (unlike my proposed—and favorite—solution 2), so perhaps this is a middle ground that you would be willing to accept, @nefeline? Definitely do let me know your thoughts.

**Reply by nefeline** [reply]
Hi there! 

> This still conflicted with the Post Content block, and I didn't dig any deeper to be honest. Again, perhaps we can change these details later to unblock this issue? In any case, this solution I have implemented still uses the native WordPress loop mechanism (unlike my proposed—and favorite—solution 2), so perhaps this is a middle ground that you would be willing to accept, @nefeline?

What do you think about following this route?

```php
                ob_start();
		rewind_posts();
		while ( have_posts() ) {
			the_post();
			comments_template();
		}
		$reviews = ob_get_clean();
```

I left this as a suggestion on https://github.com/woocommerce/woocommerce-blocks/pull/8857#discussion_r1148729004 : IMO this is a clean way of getting the Product Reviews block up running error-free and without any conflicts with the Post Content block or Featured Image blocks, and at the same time preserving hooks.

> If both blocks are present on the same page, since they do not use the rewind_posts mechanism that you proposed, wouldn't calling the_post increase the counter and make them out of sync? Then, maybe they have some other way to keep the loop from making side effects (which is my main worry)?

I think this could be potentially problematic if we were talking about the main blog loop with multiple posts: indirectly, the Gutenberg blocks are also invoking `rewind_posts();` by calling `have_posts()` (which will rewind the array if the current post == the post count). I'm also curious about potential side effects that we didn't uncover yet, although I suspect if Gutenberg is already doing the same (directly and indirectly), maybe that's what we should be doing as well?

**Reply by nefeline** [reply]
Out of curiosity, I just checked what Gutenberg does for rendering the comments block, and they [are also relying on](https://github.com/WordPress/gutenberg/blob/e9d91990d8a0ca3bf808334d8fdf68f6ed2ebe88/packages/block-library/src/comments/index.php#L60) on `comments_template()`, so that's probably the path we should follow here? What are your thoughts?

Maybe in the long term, we could also consider relying on extending that block for rendering our reviews/comments instead?

**Reply by nefeline** [reply]
For future reference: this conversation continued on the Add Product Reviews block PR, and we have decided to stick with what GB does for rendering the comments template and `rewind_posts();` : https://github.com/woocommerce/woocommerce-blocks/pull/8857#discussion_r1148729004

**Comment by gigitux** [comment]
> The reason for this is that the ”Post Content” block [calls the_post function](https://github.com/WordPress/gutenberg/blob/f7052d3e1e3e69f3b09b412c071c6c1eac69836f/packages/block-library/src/post-content/index.php#L41), which is a function with [global side-effects](https://developer.wordpress.org/reference/functions/the_post/): it doesn't just set the context, but in fact moves the index of the loop forward when it is called.

I can confirm that https://github.com/WordPress/gutenberg/pull/49904/files fixed this behavior:


| Before | After |
|--------|--------|
|![image](https://user-images.githubusercontent.com/4463174/233368227-9fd37e00-70f8-42a9-9ad0-18994423419f.png)|![image](https://user-images.githubusercontent.com/4463174/233367633-c3acacb4-07f4-4acc-bc36-2fdffb052407.png)| 

cc @ntsekouras

**Comment by passatgt** [comment]
I think it would be a good idea to have a separate product description, additional information and reviews block. And maybe a "tabs" block, if someone wants to display it that way. Bot honestly, a simple 2 column layout, left side for description, right side for reviews + attributes would be perfect, we don't need tabs anymore.

**Reply by sunyatasattva** [reply]
Hey @passatgt Reviews block is already available, the other blocks are also on the roadmap! Thank you for the feedback, that's valuable!

**Reply by passatgt** [reply]
That block is to list all reviews, it doesn't work on a single product page, no?

**Reply by sunyatasattva** [reply]
@passatg There is a difference between “All Reviews” block and “Product Reviews”. You can use the “Product Reviews” block on the Single Product template, and it will add the form and the current reviews. Try it out and let us know how it goes!
